File: apps/api/app/services/documents.py
from __future__ import annotations

import hashlib

from sqlmodel import Session

from app.models.base import utcnow
from app.models.document import Document, DocumentStatus
from app.services.document_extraction import (
    DocumentExtractionError,
    count_pdf_pages,
    detect_source_type,
    extract_text,
)
from app.storage import object_keys
from app.storage.local_storage import LocalObjectStorage


def create_document(
    session: Session,
    storage: LocalObjectStorage,
    *,
    filename: str,
    mime_type: str | None,
    data: bytes,
    title: str | None = None,
) -> Document:
    """Persist an uploaded document: store bytes, best-effort extract text + page count.

    Extraction runs at upload so non-PDF documents render immediately and PDFs get a page
    count. Indexing (Phase 3) is separate and preferentially sends original bytes to SlimX-RAG.
    """
    source_type = detect_source_type(filename, mime_type)
    content_hash = hashlib.sha256(data).hexdigest()

    doc = Document(
        title=title or filename,
        filename=filename,
        mime_type=mime_type,
        source_type=source_type,
        object_key="pending",
        content_hash=content_hash,
        file_size=len(data),
        status=DocumentStatus.UPLOADING,
    )
    session.add(doc)
    session.flush()  # assign doc.id

    original_key = object_keys.document_original_key(doc.id, filename)
    storage.put_bytes(original_key, data)
    doc.object_key = original_key

    if source_type == "pdf":
        doc.page_count = count_pdf_pages(data)

    # Best-effort text extraction — a failure here (e.g. scanned PDF) never blocks the upload;
    # the document still opens and indexing will report a precise reason later.
    try:
        _, text = extract_text(filename, mime_type, data)
    except DocumentExtractionError:
        text = ""
    if text.strip():
        text_key = object_keys.document_extracted_text_key(doc.id)
        storage.put_text(text_key, text)
        doc.extracted_text_key = text_key
        doc.status = DocumentStatus.PARSED
    else:
        doc.status = DocumentStatus.UPLOADED

    doc.updated_at = utcnow()
    session.add(doc)
    session.commit()
    session.refresh(doc)
    return doc

File: apps/web-lite/src/local-api/documents.ts
// IndexedDB-backed replacement for apps/web/lib/api/documents.ts, mirroring the upload semantics
// of apps/api/app/services/documents.py::create_document: store bytes, best-effort extraction
// (a scanned PDF never blocks the upload), then enqueue indexing.
import type { Document, DocumentContent, DocumentList } from '@web/lib/types';
import { getDb, newId, sha256Hex, utcnow } from '../db/db';
import {
  detectSourceType,
  ExtractionError,
  extractText,
  isSupportedDocument,
} from '../extraction';
import { enqueueIndexing } from '../rag/indexingController';
import { ApiError, notFound } from './errors';
import { ensureFileUrl, revokeFileUrl } from './fileUrls';

export const MAX_UPLOAD_MB = 50;

export async function listDocuments(params?: {
  q?: string;
  limit?: number;
  offset?: number;
}): Promise<DocumentList> {
  const db = await getDb();
  let docs = await db.getAll('documents');
  const q = params?.q?.trim().toLowerCase();
  if (q) {
    docs = docs.filter(
      (d) => d.title.toLowerCase().includes(q) || d.filename.toLowerCase().includes(q),
    );
  }
  docs.sort((a, b) => (a.created_at < b.created_at ? 1 : -1));
  const offset = params?.offset ?? 0;
  const limit = params?.limit ?? 50;
  return { items: docs.slice(offset, offset + limit), total: docs.length, limit, offset };
}

export async function getDocument(documentId: string): Promise<Document> {
  const db = await getDb();
  const doc = await db.get('documents', documentId);
  if (!doc) throw notFound();
  await ensureFileUrl(documentId);
  return doc;
}

export async function getDocumentContent(documentId: string): Promise<DocumentContent> {
  const db = await getDb();
  const doc = await db.get('documents', documentId);
  if (!doc) throw notFound();
  const extracted = await db.get('extracted', documentId);
  const text = extracted?.text ?? '';
  return {
    document_id: documentId,
    source_type: doc.source_type,
    text,
    available: text.trim().length > 0,
  };
}

export async function uploadDocument(file: File): Promise<Document> {
  if (!isSupportedDocument(file.name, file.type || null)) {
    throw new ApiError(`Unsupported document type: ${file.name}`, {
      status: 415,
      code: 'unsupported_type',
    });
  }
  if (file.size > MAX_UPLOAD_MB * 1024 * 1024) {
    throw new ApiError(`File exceeds the maximum upload size (${MAX_UPLOAD_MB} MB).`, {
      status: 413,
      code: 'file_too_large',
    });
  }

  const data = await file.arrayBuffer();
  const now = utcnow();
  const doc: Document = {
    id: newId(),
    title: file.name,
    filename: file.name,
    mime_type: file.type || null,
    source_type: null,
    content_hash: await sha256Hex(data),
    file_size: file.size,
    page_count: null,
    status: 'uploaded',
    indexing_status: 'not_indexed',
    rag_index_ref: null,
    metadata: {},
    created_at: now,
    updated_at: now,
  };

  // Best-effort extraction — a failure (e.g. scanned PDF) never blocks the upload; indexing will
  // report a precise reason later.
  let text = '';
  let pages = null;
  try {
    const extraction = await extractText(file.name, file.type || null, data);
    doc.source_type = extraction.sourceType;
    doc.page_count = extraction.pageCount;
    text = extraction.text;
    pages = extraction.pages;
  } catch (err) {
    doc.source_type = doc.source_type ?? null;
    if (err instanceof ExtractionError) doc.page_count = err.pageCount;
  }
  if (!doc.source_type) {
    doc.source_type = detectSourceType(file.name, file.type || null);
  }
  doc.status = text.trim() ? 'parsed' : 'uploaded';

  const db = await getDb();
  await db.put('files', {
    document_id: doc.id,
    filename: file.name,
    mime_type: file.type || null,
    data,
  });
  if (text.trim()) {
    await db.put('extracted', { document_id: doc.id, text, pages });
  }
  await db.put('documents', doc);

  const job = await enqueueIndexing(doc.id);
  return { ...doc, indexing_status: job.status as Document['indexing_status'] };
}

export async function deleteDocument(documentId: string): Promise<void> {
  const db = await getDb();
  revokeFileUrl(documentId);
  const tx = db.transaction(
    ['documents', 'files', 'extracted', 'annotations', 'notes', 'chunks', 'indexing_jobs'],
    'readwrite',
  );
  await tx.objectStore('documents').delete(documentId);
  await tx.objectStore('files').delete(documentId);
  await tx.objectStore('extracted').delete(documentId);
  for (const store of ['annotations', 'notes', 'chunks', 'indexing_jobs'] as const) {
    let cursor = await tx.objectStore(store).index('by_document').openCursor(documentId);
    while (cursor) {
      await cursor.delete();
      cursor = await cursor.continue();
    }
  }
  await tx.done;
}
